### ISSUE TITLE
Akira only opens design files when running as superuser

### ISSUE THREAD

**Issue by xoascf** [akiraux/Akira#441]
## Expected Behaviour
Akira should be able to open design files regardless of the user's permissions.

## Current Behaviour
Akira needs root permissions to access the content.json file when decompressing it from the design file (possibly by compressing it without setting permissions and preserving them when decompressing).

## Possible Solution
Change file permissions when compressing and decompressing (for older versions with this bug).

## Steps to Reproduce (for bugs)
1. Follow the instructions in the README.md file (Compile & Run section)
2. Open Akira...
3. Insert any shape
4. Save the file
5. Close Akira, then reopen
6. Try to open the previously saved file...

## Context
A design application must be fully functional when opening and saving files.
I tried to design something, save it to continue later, but the design didn't show, at first I was scared because I thought it didn't save, then I checked the log, got the permission denied error for the content.json file (ZipArchiveHandler.vala:62), tried to run Akira as superuser and it "worked".

## Your Environment (for bugs)
- Akira Version: build from master
- OS Info: Ubuntu 20.10

**Comment by albfan** [comment]
zip file do not preserve file permissions.

https://superuser.com/a/603071/158200

Probably you're running akira as root?

**Comment by xoascf** [comment]
Zip files do preserve permissions, try it and see for yourself. No, I said I tried it as root later, it happens that there are no write or read permissions on that specific file when running Akira as a standard user.

**Comment by albfan** [comment]
from man zipinfo:

```
-z     include the archive comment (if any) in the listing.
```

But looks like its possible:

https://serverfault.com/a/585889/343281

Anyway, error is at same time the solution

> Akira needs root permissions...

You run akira under root. That's the only way akira can set root permissions.

Try with to save a new draw and check what you did wrong at first time.

**Comment by xoascf** [comment]
`entry.set_perm (0644);` in ZipArchiveHandler.vala

Well, that will fix this issue, at least it did in my local repository...

**Comment by xoascf** [comment]
Yes, that fixed the loading and saving of the design file...
I will make a few more changes and prepare a pull request.

**Comment by albfan** [comment]
Yes, open a PR and we can review, but see right now you're fixing the: "it fails on my system" as we don't know the causes here.

Remember to provide examples of what you get when you save something and how it fails for you.

**Comment by xoascf** [comment]
I'm not fixing anything, I'm just assuming that the developer, when writing the code, made a typo, since this line doesn't make sense:
[`entry.set_perm (Archive.FileType.IFREG);`](https://github.com/akiraux/Akira/blob/27ce1a05e54c21fa1328d67a32d1f8163656ee4e/src/FileFormat/ZipArchiveHandler.vala#L332)

Examples:
Failed to open file (Permission denied):
```
(Akira:73881): Akira-WARNING **: 15:29:47.375: ZipArchiveHandler.vala:62: Falló al abrir el archivo «/home/xoas/.~lock.akira.Pape.akira/content.json»: Permiso denegado
```

File without permissions inside the Akira file (ZIP) saved with the [master branch](https://github.com/akiraux/Akira/tree/de83889a7880cd93170c8d9213db7218ba5114a6) build:
```
unzip -Z '/home/xoas/Pape.akira' 
Archive:  /home/xoas/Pape.akira
Zip file size: 1879 bytes, number of entries: 1
----------  2.0 unx    17338 bX defN 80-Jan-01 00:00 content.json
1 file, 17338 bytes uncompressed, 1701 bytes compressed:  90.2%
```

With permissions after PR #443:
```
unzip -Z '/home/xoas/sdf.akira'  
Archive:  /home/xoas/sdf.akira
Zip file size: 795856 bytes, number of entries: 2
-rw-r--r--  2.0 unx    10030 bX defN 80-Jan-01 00:00 content.json
-rw-r--r--  2.0 unx   803055 bX defN 80-Jan-01 00:00 Pictures/akira-img-0-1600641880.png
2 files, 813085 bytes uncompressed, 795476 bytes compressed:  2.2%
```

There are no longer any errors when opening any of the two previously saved files.

**Comment by albfan** [comment]
libarchive filetype vapi changed (don't remember exact details but filetype was assigned to set_perm or similar) and that is needed to make it work on that version.

if you see a condition about vala, avoid to remove the code if it doesn't affect you.

So what is your vala version and your libarchive version. That's your problem, and the combination that needs to be fixed.

**Comment by xoascf** [comment]
Um... ok.

libarchive: Libarchive 3.4.3
vala: Vala 0.50.0.4-69cc1

**Comment by xoascf** [comment]
This?
[libarchive: Fix Archive.[set_]mode/perm() and add Archive.FileMode ](https://gitlab.gnome.org/GNOME/vala/-/commit/b8e1a4e591fde375dc46c72d27ed31e2d06ef072)

**Comment by xoascf** [comment]
Right... error: Argument 1: Cannot convert from int' to Archive.FileType' entry.set_perm( 0644 );

**Comment by xoascf** [comment]
There seem to be only two solutions merging PR #443 with some minor nitpicks regarding the fallback version:
Either provide the fixed vapi for libarchive (assuming the user has a recent version of libarchive that has no problems), or upgrade to a similar version of Vala from 0.48 or similar as a fallback.

**Comment by albfan** [comment]
@xoascf yes, at least you just need to keep same conditional compilation. libarchive examples shows that force that permissions is needed so we are ok with those changes.

Just rework that and we are ready to merge

**Comment by xoascf** [comment]
Ok, I've compiled with Vala 0.40.23 and Libarchive 3.2.2 (elementary OS host) and no error was caused by compiling the first commit of the PR #443, it seems it doesn't affect it, anyway I'll make a new commit without the conditional compilation, so it won't affect others who want to use previous versions.

**Comment by xoascf** [comment]
Compiles the same with previous versions (Vala 0.40.23 and Libarchive 3.2.2) without problems with the updated code (without the conditional compilation).

**Comment by xoascf** [comment]
> @xoascf yes, at least you just need to keep same conditional compilation. libarchive examples shows that force that permissions is needed so we are ok with those changes.
> 
> Just rework that and we are ready to merge

Oh... Well...
If you want me to keep the conditional compilation I can revert to the previous commit, no problem, it's the same, it compiles.

**Comment by albfan** [comment]
Revisiting how vala declares their defines:

https://gitlab.gnome.org/GNOME/vala/-/blob/master/vala/valacodecontext.vala#L570

Looks like each version defines all the previous DEFINES.

Looking for the libarchive change, it was resolved on vala 0.46

```
* | | 3c5d2b605 (tag: 0.46.0) Release 0.46.0
* | | 0491b3c0e (tag: 0.45.91) Release 0.45.91
| | | * 6eba01321 (tag: 0.36.20, origin/0.36) Release 0.36.20
| * | | f9c0e566f (tag: 0.40.16) Release 0.40.16
| | * | 4f55ca372 (tag: 0.44.7) Release 0.44.7
* | | | bacc6e6d9 (tag: 0.45.90) Release 0.45.90
| | * | d7a9dece7 (tag: 0.44.6) Release 0.44.6
* | | | e3ba95e48 (tag: 0.45.3) Release 0.45.3
* | | | f2f25a6cc (tag: 0.45.2) Release 0.45.2
| | * | bc4ae5afe (tag: 0.44.5) Release 0.44.5
| | * | 27153b0cb (tag: 0.44.4) Release 0.44.4
* | | | 9365176e3 (tag: 0.45.1) Release 0.45.1
| | | | * f75cfa148 (origin/wip/traverse-visitor) Add TraverseVisitor for traversing the tree with a callback
| |_|_|/  
|/| | |   
| | | | * ecb05993e (origin/786-add-gir-version-annotation) vala: anotate valaversion file
| |_|_|/  
|/| | |   
| | | | * d5adda2f7 (origin/wip/code-transformer) tests: Add "postcondition with local result" test to increase coverage
| |_|_|/  
|/| | |   
| | | * ab544aa99 (tag: 0.36.19) Release 0.36.19
| | | | * b7db533f1 (tag: 0.42.7, origin/0.42) Release 0.42.7
| * | | | f3b20c8b7 (tag: 0.40.15) Release 0.40.15
| | * | | 2ddcfe403 (tag: 0.44.3) Release 0.44.3
* | | | | b8e1a4e59 (libarchive) libarchive: Fix Archive.[set_]mode/perm() and add Archive.FileMode
| | * | | 7a9f36448 (tag: 0.44.2) Release 0.44.2

```

#445 should tackle this.

@xoascf thanks for looking into this
